### PR TITLE
Migrate to Twitter API v2

### LINKED ISSUE #173: Check if plugin can work with only Essential access
### Is your enhancement related to a problem? Please describe.

For now, the plugin requires Elevated access to Twitter API. Elevated access is required maily for upload media which is not available under Twitter API 2.

Recently, Twitter has announced that V1.1 media endpoints will be available under Essential access. more info: https://twittercommunity.com/t/v1-1-media-endpoints-available-for-essential-access-in-the-twitter-api-v2/171664

 The user has to apply for Elevated access from the developer portal and after getting approval from the Twitter team they can use API credentials. This issue is created to check on upload media functionality and make sure all working fine with `Essential access`. If all is working as expected we can remove setup steps from settings related to requesting Elevated access, which will make user onboarding simple.

### Designs

_No response_

### Describe alternatives you've considered

Keep it as it is.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by iamdharmesh**
### Description of the Change
PR migrates the current Twitter API implementation to use API v2. Majorly PR has made the below changes
- Migrated to Twitter API v2
- Add migrate to v2 notice for only existing users
- Update app setup steps to the latest process
- Added learn more link to error message for old credentials user

Closes #173

### How to test the Change
Follow the steps mentioned on the `Autoshare for Twitter Settings`  admin page and set up the Twitter app. You should be able to set up the plugin without struggle. After saving the plugin setting with the Twitter app details, you should be able to tweet from WordPress from the post edit page. 

1. Make sure the plugin works fine with Twitter API v2
2. Verify plugin shows notice about migrating to v2 to existing users
3. Error message to existing v1.1 API users while tweet post.

### Changelog Entry

> Added - Support for Twitter API v2.
> Changed - Update plugin settings and guidelines to set up a Twitter app.

### Credits

Props @iamdharmesh, @qasumitbagthariya, @ravinderk, @jeffpaul.

## Noticeable changes
- Add PHP dependency to `7.4` in composer.json

### Checklist:

- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

**Comment by ravinderk** [comment]
@iamdharmesh For some reason, E2E tests are failing. This pull request does not change anything related to UI/UX. Do you know if these tests are falling correctly?

**Comment by iamdharmesh** [comment]
@ravinderk I have reverted app setup instructions as we discussed internally on call, Also tests seems failing due to credentials currently setup in repo secret is not migrated to Twitter API V2 yet.

Thanks

**Comment by jeffpaul** [comment]
@iamdharmesh I added the `*_V2` repo secrets and then updated the cypress workflow, but still seems to have some failures.  Could you please check those?

**Comment by jeffpaul** [comment]
Let's update the sidebar error from:
> Failed to tweet: Error: . `<Twitter error here>`

to:
> Failed to tweet; error: `<Twitter error here>`

...also, if the Twitter error is the "When authenticating requests to the Twitter API v2 endpoints, you must use keys and tokens from a Twitter developer App that is attached to a Project. You can create a project via the developer portal."  (hopefully there's an error code we can target in case they change that copy) then let's please add after the Twitter error message "Learn more here." and link to https://developer.twitter.com/en/docs/twitter-api/migrate/ready-to-migrate.

...which means in this case that would be like:
> Failed to tweet; error: `<Twitter error here>` [Learn more here.](https://developer.twitter.com/en/docs/twitter-api/migrate/ready-to-migrate)

**Comment by jeffpaul** [comment]
On the Settings page, update all references existing to match spelling/case of:
- API Key
- API Key Secret
- Access Token
- Access Token Secret

...this way we match the exact same spelling/case/wording used on https://developer.twitter.com to avoid user confusion on configuring the plugin.

**Comment by jeffpaul** [comment]
Let's update the admin notice to:
Autoshare for Twitter now utilizes the [Twitter v2 API](https://developer.twitter.com/en/products/twitter-api).  If you have not already done so, please [migrate your app](https://developer.twitter.com/en/portal/projects-and-apps) to Twitter v2 API to continue using Autoshare for Twitter.  [Learn more about migrating here](https://developer.twitter.com/en/docs/twitter-api/migrate/ready-to-migrate).

**Comment by jeffpaul** [comment]
@iamdharmesh the above couple updates should bring this PR to a good point for final review/merge... thanks!

**Comment by iamdharmesh** [comment]
> Failed to tweet: Error: . `<Twitter error here>`

@jeffpaul Currently we are also showing an error code in this message. for eg: `Error: 429. Too Many Requests`. (dot there looks extra due to no error code) So, should we keep it showing the error code OR it is no longer needed?

**Comment by jeffpaul** [comment]
If there's no code then let's not display the "Error: ." bit, if there is a code then we can keep that wrapping text.

**Comment by iamdharmesh** [comment]
>@iamdharmesh I added the *_V2 repo secrets and then updated the cypress workflow, but still seems to have some failures. Could you please check those?

It failed due to the API limit was exceeded for 24 hours.

File: autoshare-for-twitter.php [hunk 1]
```diff
@@ -41,3 +41,14 @@
...
 do_action( 'autoshare_for_twitter_loaded' );
+
+/**
+ * Register an activation hook that we can hook into.
+ */
+register_activation_hook(
+	__FILE__,
+	function () {
+		// Don't need to show migration notice to new users.
+		update_option( 'migrate_to_twitter_v2_api_notice_dismissed', true );
```
Comment (Sidsector9) [inline autoshare-for-twitter.php]: A nitpick here - will be good to prefix the option with `autoshare_`

File: includes/admin/post-meta.php [hunk 2]
```diff
@@ -317,7 +317,7 @@ function get_tweet_status_message( $post ) {
 
 			case 'error':
 				$response_array[] = [
-					'message' => __( 'Failed to tweet: ', 'autoshare-for-twitter' ) . $tweet_meta['message'],
+					'message' => __( 'Failed to tweet; ', 'autoshare-for-twitter' ) . $tweet_meta['message'],
```
Comment (Sidsector9) [inline includes/admin/post-meta.php:320]: Is this intentional? Colon `:` would be appropriate here.

**Review by Sidsector9 — CHANGES_REQUESTED** [review]
Thanks for this PR! I've requested some minor changes, after that it's good to go 🎉

File: includes/admin/post-meta.php (hunk 2 above)
Comment (iamdharmesh) [reply includes/admin/post-meta.php:320]: @Sidsector9 it is updated as per the request here https://github.com/10up/autoshare-for-twitter/pull/229#issuecomment-1544198357

File: includes/admin/post-meta.php [hunk 3]
```diff
@@ -434,11 +434,19 @@ function markup_published( $status_meta ) {
  * @return string
  */
 function markup_error( $status_meta ) {
+	$learn_more = '';
+	if ( 'When authenticating requests to the Twitter API v2 endpoints, you must use keys and tokens from a Twitter developer App that is attached to a Project. You can create a project via the developer portal.' === $status_meta['message'] ) {
```
Comment (ravinderk) [inline includes/admin/post-meta.php:438]: @iamdharmesh, is this message coming from Twitter? Do you know if Twitter responds in English always?
I want to ensure that it succeeds Twitter below to a different language.

File: src/js/AutoshareForTwitterPostStatusInfo.js [hunk 4]
```diff
@@ -8,6 +8,17 @@ import { useHasFeaturedImage, useAllowTweetImage, useSaveTwitterData } from './h
 
 import { getIconByStatus } from './utils';
 
+// Error message component.
+const ErrorMessage = ( { errorMessage } ) => {
+	return (
+		<span>
+			{ errorMessage }
+			{ ' ' }
+			{ ( errorMessage === 'Failed to tweet; When authenticating requests to the Twitter API v2 endpoints, you must use keys and tokens from a Twitter developer App that is attached to a Project. You can create a project via the developer portal.' ) && <ExternalLink href={ 'https://developer.twitter.com/en/docs/twitter-api/migrate/ready-to-migrate' }>{ __( 'Learn more here.' ) }</ExternalLink> }
```
Comment (ravinderk) [inline src/js/AutoshareForTwitterPostStatusInfo.js]: @iamdharmesh, did we leave it intentionally because this message is not translatable?

Comment (ravinderk) [reply src/js/AutoshareForTwitterPostStatusInfo.js]: I think it is related to https://github.com/10up/autoshare-for-twitter/pull/229/files/cfcf79d1dbd7e82ef2ac3541f19a98a86cf8419c..cbcfab22b9f84ba313dec5a5c8a1285153639624#r1194775171

**Review by ravinderk — COMMENTED** [review]
@iamdharmesh I left a few questions.

File: includes/admin/post-meta.php (hunk 3 above)
Comment (iamdharmesh) [reply includes/admin/post-meta.php:438]: Yes @ravinderk, this message coming from Twitter. I tried to check but didn't found anywhere that twitter respond error message other than English language.

File: src/js/AutoshareForTwitterPostStatusInfo.js (hunk 4 above)
Comment (iamdharmesh) [reply src/js/AutoshareForTwitterPostStatusInfo.js]: Yes @ravinderk, as this message comes from Twitter API. we just compare here it identify the error and append learn more link to error message.

**Review by Sidsector9 — APPROVED** [review]

**Pull request merged**